### ISSUE TITLE
[Bug]: NVFP4 garbage output under torch.compile when per-rank K is not 32-aligned (TP8 on SM120): padded block-scale column reads stale memory; eager/Marlin/aligned-TP correct

### ISSUE THREAD

**Issue by jszzr** [vllm-project/vllm#48898]
### Your current environment

<details>
<summary>Environment</summary>

```
vLLM: 0.25.1 (pip install, official wheel)
PyTorch: 2.11.0+cu130
flashinfer-python: 0.6.13, flashinfer-cubin: 0.6.13 (no flashinfer-jit-cache installed)
transformers: 5.14.1
OS: Ubuntu 22.04.4, kernel 5.15.0-1069-nvidia, glibc 2.35
Python: 3.10.12
GPU: 8x NVIDIA RTX 6000D (Blackwell workstation, compute capability 12.0 / SM120, ~84 GB each)
Driver: 580.95.05, CUDA 13.0
CPU: 2x AMD EPYC 9575F
```
</details>

### 🐛 Describe the bug

Serving a **ModelOpt NVFP4** checkpoint of a **dense** model (Qwen2.5-VL-72B-Instruct, quantized with nvidia-modelopt 0.45.0, unified HF export — same quant config as `nvidia/Qwen2.5-VL-7B-Instruct-NVFP4`: NVFP4, group_size 16, KV cache unquantized, `exclude_modules: [lm_head, model.language_model.embed_tokens, model.visual*]`) on **SM120** produces **completely broken output** with the auto-selected NVFP4 linear kernel, and still-incorrect output with the native CUTLASS kernel. Only Marlin produces correct results.

The checkpoint itself is fine: modelopt's own post-PTQ generation check is coherent, and vLLM with `--linear-backend marlin` is fully correct.

Serve command:

```bash
vllm serve Qwen2.5-VL-72B-Instruct-NVFP4 --quantization modelopt_fp4 \
    --tensor-parallel-size 8 --max-model-len 16384 --port 8000
    # plus --linear-backend {cutlass|marlin} for the variants below
```

Results per NVFP4 linear kernel (greedy decoding, `temperature=0`):

| `--linear-backend` | Kernel (log line) | `"What is 15 + 27?"` (chat) | `"The capital of France is"` (raw completion) | GSM8K (300, flexible-extract) | MMMU val (300, single-image MC) |
|---|---|---|---|---|---|
| auto | `FlashInferCutlassNvFp4LinearKernel` | `'/Dk'` (byte garbage) | — | **0.0** | **0.07** |
| `cutlass` | `CutlassNvFp4LinearKernel` | fluent but unrelated text (hallucinates an image description) | `'-Zafer Yilmaz\nThe most'` | — | — |
| `marlin` | `MarlinNvFp4LinearKernel` | `'42'` ✅ | `' Paris. The capital of Germany is Berlin'` ✅ | correct (full eval in progress; spot checks all pass) | — |
| (reference) BF16 original model | — | correct | correct | 0.887 | 0.69 |

Two distinct problems:

1. **`FlashInferCutlassNvFp4LinearKernel` is auto-selected on SM120 but is known-broken there.** Its `is_supported()` only checks `has_device_capability(100)`, so SM120 passes. But FlashInfer's `mm_fp4(backend="cutlass")` on SM120 has documented correctness bugs: flashinfer-ai/flashinfer#2577 (still open: cutlass backend returns zeros on RTX PRO 6000), flashinfer-ai/flashinfer#2708 (missing GDC flags → PDL barriers no-op → NaN/garbage tiles under concurrency). SGLang explicitly **gates flashinfer-cutlass FP4 off on SM120** (sgl-project/sglang#18954, fixed by sgl-project/sglang PR #20047 which defaults SM120 to cudnn); vLLM 0.25.1 has no equivalent gate. Suggested fix: exclude SM120 in `FlashInferCutlassNvFp4LinearKernel.is_supported()` (or rank it below the native kernels there).

2. **Native `CutlassNvFp4LinearKernel` (`cutlass_scaled_fp4_mm_sm120a`) also produces incorrect results on SM120** in this setup — output is locally fluent but semantically wrong even for a 5-token greedy completion, which points at numerically corrupted GEMM rather than sampling issues. This one looks like a separate bug in the sm120a path (or its interaction with TP=8).

Happy to run further diagnostics on this machine (8x SM120) — e.g. single-GPU repro with a small NVFP4 model, kernel unit tests, or a patched wheel.

**Workaround for anyone hitting this:** `--linear-backend marlin` gives correct output on SM120 (at reduced prefill throughput vs native FP4).

**Comment by jszzr** [comment]
Follow-up with full eval numbers for the Marlin workaround (same setup as above, TP8, greedy):

| Metric | BF16 original | NVFP4 `--linear-backend marlin` | NVFP4 auto (FlashInferCutlass, broken) |
|---|---|---|---|
| GSM8K (300, flexible-extract) | 0.8867 | **0.8867** | 0.0 |
| MMMU val (300, single-image MC) | 0.690 | **0.683** | 0.07 |
| TPOT @ concurrency 1 (1024 in / 256 out) | 24.30 ms | **14.94 ms** (1.63x) | 15.10 ms |
| Total token throughput @ concurrency 64 | 3207 tok/s | 3380 tok/s | 4289 tok/s |

So on SM120, Marlin fully preserves quality and matches the (numerically broken) FlashInfer-cutlass path on single-stream decode latency; the remaining gap is batch throughput, where the broken native-FP4 path was ~27% ahead of Marlin — that's the upside of fixing the CUTLASS kernels on SM120 for dense models.

**Comment by jszzr** [comment]
Additional diagnostics — ruling out the local-JIT-compilation hypothesis and adding `flashinfer_cudnn` to the broken list.

We installed the official prebuilt `flashinfer-jit-cache==0.6.13+cu130` wheel (sha256 verified against the flashinfer.ai index) and re-tested. **No change**: prebuilt cubins produce the same garbage as the locally-JIT-compiled path.

Full backend matrix on SM120 (8x RTX 6000D, TP8, ModelOpt NVFP4 dense Qwen2.5-VL-72B, greedy):

| `--linear-backend` | jit-cache absent | jit-cache 0.6.13 installed |
|---|---|---|
| `flashinfer_cutlass` (auto-selected) | byte garbage (`'/Dk'`) | still garbage (emits C code / hallucinated image descriptions) |
| `flashinfer_cudnn` (SGLang's SM120 default) | garbage (wrong-language ramblings) | still garbage |
| `cutlass` (native, prebuilt in wheel) | fluent but semantically wrong | n/a (not flashinfer) |
| `marlin` | ✅ correct (GSM8K 0.8867 == BF16) | ✅ correct |

Observations that may help localize the bug:
- Three independent GEMM implementations (flashinfer-cutlass, flashinfer-cudnn, vLLM native cutlass sm120a) all produce incorrect results, while Marlin — which does its own weight repacking — is exact. That pattern is consistent with a problem in the **shared NVFP4 weight/scale preprocessing path** (e.g. the swizzled block-scale layout fed to the "native FP4" kernels) on SM120 rather than three separate kernel bugs — though flashinfer#2577 suggests at least the flashinfer-cutlass path also has genuine kernel-level SM120 issues.
- Failure modes differ by backend (byte garbage vs fluent-but-wrong), so corruption severity varies with the kernel consuming the same preprocessed weights.

Environment as in the original report, plus: cuDNN 9.19 (bundled nvidia-cudnn-cu13), flashinfer-python 0.6.13, flashinfer-cubin 0.6.13, flashinfer-jit-cache 0.6.13+cu130.

**Comment by jszzr** [comment]
Root cause isolated. The garbage output is **not** a kernel numerics bug and **not** SM120-specific: it is an unsound zero-initialization assumption in the NVFP4 activation-quant **K-padding path** that torch.compile/inductor's buffer reuse violates. Full verdict matrix on the same machine/checkpoint/stack as the original report (vLLM 0.25.1, flashinfer 0.6.13, 8× RTX 6000D):

| TP | K_rank (down_proj) | 32-aligned | mode | verdict | GSM8K (50, flex) |
|---|---|---|---|---|---|
| 1 | 29568 | yes | compiled | COHERENT | 0.66 |
| 2 (both PCIe topologies) | 14784 | yes | compiled | COHERENT | — |
| 4 | 7392 | yes | compiled | COHERENT | — |
| **8** | **3696** | **NO** | **compiled** | **GARBAGE** | **0.00** |
| 8 | 3696 | NO | compiled, PDL off | GARBAGE (PDL exonerated) | — |
| **8** | **3696** | **NO** | **eager** | **COHERENT** | **0.64** |
| 1 (marlin) | — | — | compiled | COHERENT | 0.58 |

## Mechanism

TP=8 is the only degree for this model where the row-parallel per-rank K
(29568/8 = 3696) is not 32-aligned, so the FP4 GEMM path pads K to 3712.
The GEMM consumes 232 block-scale columns per row, but the activation-quant
kernels (`scaled_fp4_quant`, fused `silu_and_mul_nvfp4_quant`) only write
231 — correctness of the padded column rests entirely on the scale buffer
being zero-initialized (the code says so itself:
`_custom_ops.py::create_fp4_scale_tensor` — "Must be zero-initialized ...
the quant kernel [does not write the padding]").

In eager mode `torch.zeros` runs per call → padding contributes exactly 0 →
TP8 output is coherent. Under torch.compile, inductor's buffer planning
treats the quant op as fully initializing its output and recycles arena
memory, so the padded scale column carries **stale bytes**; any byte that
decodes to a large/NaN FP8 poisons every output element. This explains all
original observations: only TP8 breaks (the only misaligned-K degree);
FlashInferCutlass and native Cutlass fail with different symptoms (they pad
by different mechanisms — post-quant `F.pad` vs in-kernel `padded_n`);
Marlin is always correct (W4A16 — no activation quant, no padding); and
isolated kernel/op tests all pass (fresh zeroed buffers).

Also verified along the way (all exonerated for THIS bug): op-level
cutlass/flashinfer NVFP4 GEMMs across M=1..1024 including odd M, layer-level
modelopt weight plumbing incl. simulated TP4/TP8 shard shapes, activation
contiguity in-engine (probe: 0 non-contiguous tensors at any TP), autotune
on/off, jit-cache present/absent, PDL on/off, PCIe topology (PIX vs NODE).

## Suggested fix

Make the FP4 quant kernels zero the padded tail of the block-scale output
themselves (both the `.out` variant and `silu_and_mul_nvfp4_quant`), or
explicitly zero-fill the pad columns at the consumer. Relying on
allocation-time zeros is unsound under inductor.

## Workarounds today

`--enforce-eager`, or `--linear-backend marlin`, or choose a TP degree whose
per-rank K stays 32-aligned.

## Secondary findings (separate PRs/issues)

1. `scaled_fp4_quant` silently quantizes wrong memory for non-contiguous
   input (Python wrapper reshapes a strided view; csrc reads linearly; no
   stride check). Not this bug's cause (in-engine activations are
   contiguous), but a latent silent-corruption hazard with a minimal repro —
   defensive fix + regression test in #49092.
2. `FlashInferTrtllmNvFp4LinearKernel.is_supported()` returns True on SM120
   but flashinfer ≥0.6.12 hard-rejects trtllm at capability 120 → guaranteed
   crash if forced; capability gate in #49093.
3. flashinfer 0.6.13 `mm_fp4(backend="auto")` raises TypeError outside
   vLLM's wrapper on SM120.
4. Flaky silent EngineCore SIGKILLs (~3/8 starts) clustered around
   flashinfer fp4_gemm autotune profiling on SM120; retries succeed.
